[Step2] 문자열 계산기

leeyohan93 opened the pull request:
리뷰어님 안녕하세요 :)
코틀린을 통해 TDD 및 클린코드를를 접하는 것은 처음이라 여러 방법을 섞어서 구현하게 되었습니다.
아래는 TODO 리스트를 통해 구현하면서 궁금했던 부분인데요! 
리뷰어님의 생각도 함께 공유받고 싶습니다 :)
감사합니다 🙇 

제가 질문을 많이 드려 글로 작성하시기 힘드시다면 연락주셔서 게더나 다른 방식으로 이야기해주셔도 좋을 것 같습니다 😄 


- Kotest를 경험하신 적이 있으시다면 자주 사용하는 스타일과 그 스타일을 사용하게 되는 이유가 궁금합니다
- 주 생성자에서 객체 생성에 대한 검증이 필요한 경우 어떤 방법이 좋을까요?
- 어떤 때에 때 코틀린의 메서드를 식으로 표현하는 것이 적절할까요?
- 어떤 때에 때 확장 함수를 사용하는 것이 적절할까요?

leeyohan93 committed:
test: Kotest에 대한 학습테스트 진행

leeyohan93 committed:
docs: 문자열 계산기에 대한 기능 요구 사항과 도메인 모델 설계 작성

leeyohan93 committed:
feat: 연산자(Operator)에 대한 테스트와 구현을 작성

leeyohan93 committed:
docs: 문자열 계산기에 대한 도메인 설계 수정

leeyohan93 committed:
feat: 수식(Expression)에 대한 테스트와 구현을 작성

leeyohan93 committed:
feat: 수식 생성(ExpressionFactory)에 대한 테스트와 구현 작성

leeyohan93 committed:
feat: 수식 문자열 분리 전략(SeparationStrategy)에 대한 테스트와 구현 작성

leeyohan93 committed:
refactor: 수식 생성(ExpressionFactory) 문자열 분리 책임 분리

leeyohan93 committed:
feat: 문자열 계산기(StringCalculator) 테스트와 구현 작성

leeyohan93 committed:
chore: 테스트 시 result 변수 명 대신 actual 사용

leeyohan93 committed:
feat: 문자열 사칙 연산 계산기에 대한 메인 함수 구현

leeyohan93 committed:
refactor: klintCheck 반영

BeokBeok commented on src/main/kotlin/calculator/CalculatorApp.kt:
```diff
@@ -0,0 +1,13 @@
+fun main() {
+    val stringCalculator = StringCalculator(ExpressionFactory(OneSpaceSeparationStrategy()))
+
+    // 입력
+    println("수식을 입력해주세요.")
+    val readText = readln()
+
+    // 출력
+    val result = stringCalculator.calculate(readText)
+    println("계산 결과: $result")
```
어디까지가 입력이고 어디까지가 출력인지 주석에 의존하고 있는 상황이네요.
입력과 출력에 대한 함수를 정의해서 분리해보면 어떨까요?

BeokBeok commented on src/main/kotlin/calculator/Expression.kt:
```diff
@@ -0,0 +1,19 @@
+    private val numbers: List<Int>,
+    private val operators: List<Operator>
+) {
+    init {
+        // TODO: Q. 주 생성자에서 객체 생성에 대한 검증이 필요한 경우 어떤 방법이 좋을까요?
+        require(numbers.size == operators.size + 1) {
+            "숫자와 연산자의 개수가 적절하지 않습니다."
+        }
+    }
```
> Q. 주 생성자에서 객체 생성에 대한 검증이 필요한 경우 어떤 방법이 좋을까요?

require문으로 검증을 잘 해주셨네요 💪 
혹시 어떤 부분을 더 고려하고 싶으신건가요?

leeyohan93 replied:
자바에서 객체 생성에 대한 검증이 필요할 때 주 생성자에서 동작하도록 구현해왔었는데, '코틀린에서는 어떻게 주 생성자에 검증 코드를 구현할 수 있을까'에 대해 궁금하여 드렸던 질문이었습니다 :)
이후에 해당 코틀린 코드를 자바로 디컴파일해보니 주 생성자에서 검증을 하도록 되어있었네요 😅 
리뷰어님도 객체 생성에 검증이 필요할 때 init을 자주 사용하는 편이신가요? 😄

BeokBeok replied:
네. init 블록에서 주로 객체 유효성 체크를 하고 있습니다 😄

BeokBeok commented on src/main/kotlin/calculator/Expression.kt:
```diff
@@ -0,0 +1,19 @@
+    }
+
+    fun calculate(): Int {
+        return numbers.reduceIndexed { index, first, second ->
+            operators[index - 1].calculate(first, second)
+        }
```
reduce 사용 👍

BeokBeok commented on src/main/kotlin/calculator/ExpressionFactory.kt:
```diff
@@ -0,0 +1,45 @@
+package calculator
+
+class ExpressionFactory(
+    private val separationStrategy: SeparationStrategy
+) {
```
전략 패턴 활용 👍

BeokBeok commented on src/main/kotlin/calculator/ExpressionFactory.kt:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    private fun numbers(texts: List<String>): List<Int> {
+        return (texts.indices step 2).map {
```
숫자 2는 어떤 것을 의미하는지 잘 모르겠습니다.
어떻게 하면 이 숫자의 의미를 잘 나타낼 수 있을까요? (이하 관련 내용 동일)

leeyohan93 replied:
매직넘버는 분리하도록 할게요 👍

BeokBeok commented on src/main/kotlin/calculator/ExpressionFactory.kt:
```diff
@@ -0,0 +1,45 @@
+        }
+    }
+
+    // TODO: 어떤 때에 때 코틀린의 메서드를 식으로 표현하는 것이 적절할까요?
+    private fun toInt(numberText: String) =
+        try {
+            numberText.toInt()
+        } catch (e: NumberFormatException) {
+            throw IllegalArgumentException("올바르지 않은 숫자를 입력하였습니다. ($numberText)")
+        }
```
> 어떤 때에 때 코틀린의 메서드를 식으로 표현하는 것이 적절할까요?

어떤 때라는 것이 구체적으로 어떤 상황이 있을까요?

leeyohan93 replied:
구체적으로 어떤 상황인지 다양하게 있지는 않지만, 
가독성에 대한 이유로 메서드를 식으로 표현할때는 한 줄 정도의 간단한 로직일 때만 사용을 해왔던 것 같습니다.
리뷰어 님은 메서드를 식으로 구현하시는 어떤 기준이 있으신지 궁금합니다 :)

BeokBeok replied:
함수에 리턴대신 `=` 이것을 넣는 것을 말씀하시는 것이였군요.
저는 예전에는 지금과 같이 간단한 return문인 경우에는 `=`를 활용했었습니다만, breakpoint를 찍어서 디버깅할 때 조금 불편한 경험이 있어서 잘 사용하지 않게 되었어요.

BeokBeok commented on src/main/kotlin/calculator/ExpressionFactory.kt:
```diff
@@ -0,0 +1,45 @@
+            throw IllegalArgumentException("올바르지 않은 숫자를 입력하였습니다. ($numberText)")
+        }
+
+    // TODO: 어떤 때에 때 확장 함수를 사용하는 것이 적절할까요?
+    private fun String.toOperator(): Operator {
+        try {
+            return Operator.of(this)
+        } catch (e: IllegalArgumentException) {
+            throw IllegalArgumentException("올바르지 않은 연산자를 입력하였습니다. ($this)")
+        }
+    }
```
> 어떤 때에 때 확장 함수를 사용하는 것이 적절할까요?

이 함수가 확장 함수로 정의가 된 것은 아마도 추출 기능을 사용해서 그런 것이라 추측되네요.
확장함수는 주로 표준 라이브러리에 없는 알고리즘을 구현해야하는 상황이고 이 알고리즘을 자주 사용하는 경우에 사용하는 것이 좋습니다.

leeyohan93 replied:
> 확장함수는 주로 표준 라이브러리에 없는 알고리즘을 구현해야하는 상황이고 이 알고리즘을 자주 사용하는 경우에 사용하는 것이 좋습니다.

공유 감사합니다 :)
그렇다면 혹시 확장함수를 private으로 구현하여 클래스 내부에서 단독적으로 사용하거나, 가독성을 위해서 사용하는 부분은 지양하시는 편이신가요?

BeokBeok replied:
어쩔수 없이 가독성을 위해서 분리하는 경우가 가끔있지만, 그 외에는 private 확장 함수는 지양하는 편입니다.

BeokBeok commented on src/main/kotlin/calculator/Operator.kt:
```diff
@@ -0,0 +1,25 @@
+
+enum class Operator(
+    private val symbol: String,
+    private val function: BiFunction<Int, Int, Int>
```
```suggestion
    private val function: (Int, Int) -> Int
```
람다를 사용하게 되면, 좀 더 코틀린스럽게 작성할 수 있어요.

BeokBeok commented on src/main/kotlin/calculator/Operator.kt:
```diff
@@ -0,0 +1,25 @@
+
+import java.util.function.BiFunction
+
+enum class Operator(
```
enum 활용 👍

BeokBeok commented on src/main/kotlin/calculator/Operator.kt:
```diff
@@ -0,0 +1,25 @@
+    ADDITION("+", { first, second -> first + second }),
+    SUBTRACTION("-", { first, second -> first - second }),
+    MULTIPLICATION("*", { first, second -> first * second }),
+    DIVISION("/", { first, second -> first / second });
```
만약에 피연산자에 0이 포함된 경우에는 어떻게 될까요?

BeokBeok commented on src/main/kotlin/calculator/Operator.kt:
```diff
@@ -0,0 +1,25 @@
+        fun of(symbol: String): Operator {
+            return Operator.values()
+                .find { it.symbol == symbol }
+                ?: throw java.lang.IllegalArgumentException("{$symbol}에 일치하는 연산자가 존재하지 않습니다.")
```
```suggestion
                ?: IllegalArgumentException("{$symbol}에 일치하는 연산자가 존재하지 않습니다.")
```
코틀린의 IllegalArgumentException을 활용할 수 있어요.

BeokBeok commented on src/main/kotlin/calculator/StringCalculator.kt:
```diff
@@ -0,0 +1,10 @@
+class StringCalculator(
+    private val expressionFactory: ExpressionFactory
+) {
+    fun calculate(expressionText: String): Int {
+        val expression = expressionFactory.create(expressionText)
+        return expression.calculate()
+    }
```
함수를 보았을 때 calculate 즉 계산한다는 의미로 보입니다.
그런데 함수 내부에는 expression을 만들어주는 로직이 들어있네요.

저는 calculate함수에서 expression을 만들 것이라 예상이 잘 안되네요.
즉 하나의 함수에서 두 가지 이상의 일을 하고 있는데, 하나의 함수에서는 하나의 일만 하도록 하면 어떨까요?

BeokBeok commented on src/test/kotlin/calculator/ExpressionFactoryTest.kt:
```diff
@@ -0,0 +1,42 @@
+    val expressionFactory = ExpressionFactory(
+        object : SeparationStrategy {
+            override fun separate(input: String): List<String> {
+                return input.split(" ")
```
```suggestion
                return input.split(OneSpaceSeparationStrategy.SEPARATOR)
```
OneSpaceSeprationStrategy에 정의한 SEPARATOR를 활용해보면 어떨까요?

leeyohan93 replied:
`OneSpaceSeprationStrategy`에 변경이 생기면 해당 테스트가 영향을 받는 상황이 생길텐데 의존성을 갖지 않는 형태로 구현하는 것이 괜찮지 않을까요? 🤔  
테스트를 구현할 때 꼭 필요한 경우가 아니라면 다른 클래스에 의존하지 않도록 하는 것을 의도하였었습니다 :)

BeokBeok replied:
가급적이면 실제 로직의 변경이 테스트 코드에 영향이 가도록 작성하는 것이 좋습니다.
이렇게 하면 프로덕트 코드 변경으로 인한 영향도를 쉽게 파악할 수 있습니다. (여기를 변경했더니 관련 테스트코드가 이것이고 이 테스트가 깨지는구나)

가령 빈 문자열을 두 칸으로 변경하게 된다면, 실제 코드와 테스트 코드 둘 다 변경해야할 것 같은데요.
만약 빈 문자열이 하나든 두개든 모두 하나로 취급하는 로직이 적용되어 있는 상태에서, 프로덕션 코드에는 빈 문자열이 두칸으로 되어 있고 테스트 코드에는 빈 문자열이 한칸으로 되어 있다면, 실제 프로덕트 코드와 테스트 코드에 적용되어 있는 seperator가 다른데 테스트 코드는 통과하지 않을까요? 🤔

leeyohan93 replied:
> 가급적이면 실제 로직의 변경이 테스트 코드에 영향이 가도록 작성하는 것이 좋습니다.
이렇게 하면 프로덕트 코드 변경으로 인한 영향도를 쉽게 파악할 수 있습니다. 

리뷰어님이 이야기해주신 이 부분이 많이 공감이 되는 것 같습니다 :)
다만 전략패턴으로 분리해놓은 부분의 구현체에서 코드의 변경이 생겼을 때, 전략 패턴을 사용하는 클래스의 테스트까지 영향을 미치도록 하는 것이 좋을지에 대해서는 조금 고민이 되는 것 같습니다.

ExpressionFactory에서 전략 패턴을 사용하여 인터페이스로 분리하여 의존성을 분리하였는데 구현체를 테스트에서 사용하게 될 경우 다시 ExpressionFactory가 전략 패턴의 구현체에 의존하는 상황이 된다고 생각하는데요.

문자열 분리 정책의 변경과 같은 코드 변경으로 인한 영향도는 전략 패턴의 구현체의 테스트(OneSpaceSeparationStrategyTest)로 파악하고, ExpressionFactory에서는 전략패턴에 의존하지 않고 그 이외의 부분을 담당하게 하게 하면 더 작은 단위테스트로 구현될 수 있지 않을까요?

> 만약 빈 문자열이 하나든 두개든 모두 하나로 취급하는 로직이 적용되어 있는 상태에서, 프로덕션 코드에는 빈 문자열이 두칸으로 되어 있고 테스트 코드에는 빈 문자열이 한칸으로 되어 있다면, 실제 프로덕트 코드와 테스트 코드에 적용되어 있는 seperator가 다른데 테스트 코드는 통과하지 않을까요? 🤔

개인적인 생각으로 이야기해주신 사례에서보면 프로덕션 코드와 테스트 코드의 seperator가 다르다는 부분보다 더욱 중요하다고 생각되는 부분은 문자열이 하나든 두개든 문자열을 올바르게 분리하고 있다는 사실인 것 같습니다. 그렇기 때문에 각 전략패턴에 대한 테스트가 중요하며, ExpressionFacotryTest 는 특정 전략 패턴에 의존하지 않고 테스트할 수 있도록 만들어야한다고 생각합니다!

BeokBeok commented on src/test/kotlin/calculator/ExpressionFactoryTest.kt:
```diff
@@ -0,0 +1,42 @@
+        val parameters = listOf(
+            "2 + 3 /",
+            "2+* 4",
+            "",
```
빈 문자열을 올바르지 않은 문자열이라고 정의하기는 애매해보여요.

leeyohan93 replied:
테스트의 이름을 적절하게 작성하지 못하여 생긴 문제 같아 보이기도 합니다 🤔 
'올바르지 않은 수식의 문자열 입력 시 예외가 발생한다' 라고 정의하면 괜찮을 것 같은데 어떨까요 ?

BeokBeok commented on src/test/kotlin/calculator/ExpressionTest.kt:
```diff
@@ -0,0 +1,38 @@
+        }
+    }
+
+    "식을 계산할 수 있다" {
```
어떤 식을 계산할 수 있는지 좀 더 명확하게 네이밍을 하면, 테스트코드가 어떤 것을 검증하는지 이해하는데 큰 도움이 될 것 같아요.

BeokBeok commented on src/test/kotlin/calculator/ExpressionTest.kt:
```diff
@@ -0,0 +1,38 @@
+
+    "식을 계산할 수 있다" {
+        val parameters = listOf(
+            Triple(listOf(2, 4), listOf(Operator.SUBTRACTION), -2),
+            Triple(listOf(2, 4, 5, 6), listOf(Operator.ADDITION, Operator.SUBTRACTION, Operator.MULTIPLICATION), 6),
+            Triple(listOf(200, 4, 5, 6), listOf(Operator.DIVISION, Operator.DIVISION, Operator.SUBTRACTION), 4),
```
Pair 혹은 Triple은 각각의 값이 어떤 것을 의미하는지 알기 어려워요.
data class를 활용하거나 혹은 각각 의미있는 변수로 정의해서 활용하면 어떨까요? (이하 관련 내용 동일)

BeokBeok reviewed (changes requested):
문자열 계산기 미션하시느라 고생하셨습니다. 👍 
고민해볼만한 의견들을 코맨트로 작성하였으니, 충분히 고민해보시고 도전해보세요. 💪

BeokBeok commented:
> Kotest를 경험하신 적이 있으시다면 자주 사용하는 스타일과 그 스타일을 사용하게 되는 이유가 궁금합니다

현업에서는 JUnit을 사용하고 있어서, 아직 Kotest는 경험이 전무합니다 🙏 
잘 부탁드립니다 😄

leeyohan93 committed:
refactor: 리뷰 피드백 반영

- 메인 메서드에서의 입출력 메서드 분리
- 계산기 수식 생성 로직 분리
- 매직넘버 분리
- Bifunction -> 람다 사용
- 나눗셈 연산 시 피연산자 0 에 대한 예외처리
- 코틀린 IllegalArgumentException 사용
- 테스트 네이밍 수정
- 테스트 픽스쳐 정의

leeyohan93 committed:
refactor: klintCheck 반영

BeokBeok commented on src/main/kotlin/calculator/StringCalculator.kt:
```diff
@@ -0,0 +1,11 @@
+class StringCalculator(
+    private val expressionFactory: ExpressionFactory
+) {
+    fun calculate(expressionText: String): Int {
+        val expression = expression(expressionText)
+        return expression.calculate()
+    }
+    private fun expression(expressionText: String) = expressionFactory.create(expressionText)
```
calculate 함수는 주어진 데이터를 가지고 계산만 하면 되는데, 내부에서 Expression으로 변환하는 로직이 들어가있었어요.
해서, 제가 의도한 것은 calculate 함수에 파라미터를 String으로 받고 있는데, 이 것을 Expression 타입으로 받게 하는 것이였어요.

leeyohan93 replied:
이 부분은 제가 리뷰어님과 생각이 조금 다르게 생각이 드는 것 같은데요 🤔 
문자열 계산기의 책임은 `문자열 값에 따라 사칙 연산을 수행할 수 있다`이라고 생각했기 때문에, 
문자열 값을 계산할 수 있는 식으로 만드는 것과 계산하는 것은 같은 추상화 수준을 가지고 함께 수행되어야 하는 책임이라고 생각이 들었습니다. 추가적으로 계산기는 Expression의 존재를 외부에 노출할 필요도 없지 않을까 하는 생각이 드는 것 같습니다!

외부에서 Expression을 만들고 calculate 함수에 파라미터로 넘겨주게 구현하게 된다면, StringCalculator는 Expression에 대한 실행 후 반환하는 로직이 구현될 것 같은데 외부에서 expression.calculate()를 하는 것과 어떤 차이가 있을까요?

BeokBeok replied:
좋은 말씀 감사합니다 😄 

저의 최종 의도는, Expression 객체에는 단순히 표현식에 대해서만 책임을 지고, StringCalculator에는 단순히 계산에 대한 책임을 지는 것이였어요.
문자열 계산기에 대한 책임에 대해서 말씀해주셨는데, 실제 계산 로직을 Expression에서 하고 있어서 저는 오히려 계산 관련 책임을 Expression에서 가지고 있는 것 같고, StringCalculator가 컨트롤러의 역할을 하고 있다고 느꼈어요. 🤔 

StringCalculator가 아닌 main 함수를 컨트롤러로 바라보신다면 어떨까요?
main함수에서 입력한 수식에 대해서 Expression을 만들어주는 함수를 호출하고, 이 Expression을 StringCalculator의 calculate 함수를 호출하면 어떨까 싶습니다.

```kotlin
private fun print(stringCalculator: StringCalculator, readText: String) {
    val expression = ExpressionFactory(OneSpaceSeparationStrategy()).create(readText)
    val result = stringCalculator.calculate(expression)
    println("계산 결과: $result")
}
```
CalculatorApp 파일에서 print함수를 제가 의도한대로 작성해봤어요. 😅 

혹시 잘 이해가 안되시면 말씀주세요!

BeokBeok commented on src/main/kotlin/calculator/ExpressionFactory.kt:
```diff
@@ -0,0 +1,49 @@
+        return Expression(numbers, operators)
+    }
+
+    private fun numbers(texts: List<String>): List<Int> {
```
private 함수들(numbers, operators, toInt 등)을 어떻게 테스트하면 좋을지 고민해보면 좋을 것 같아요.

BeokBeok commented on src/main/kotlin/calculator/ExpressionFactory.kt:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    companion object{
+        const val TEXT_INTERVAL = 2
```
내부에서만 사용하는 경우에는 반드시 private 접근자를 붙여주세요.

BeokBeok replied:
TEXT_INTERVAL이라는 명칭이 저는 조금 이해가 잘 안가네요.
TEXT_INTERVAL은 어떤 역할을 하나요?

leeyohan93 replied:
TEXT_INTERVAL이 숫자나 연산자의 인덱스간격이 2이기 때문에 설정한 이름인데요!
뭔가 더욱 적합한 이름이 떠오르지 않네요 😂  

지금 로직에서는 조금 표현하기 어려운 느낌이라 구현 로직을 리팩토링해보겠습니다 :)

BeokBeok commented on src/test/kotlin/calculator/StringCalculatorTest.kt:
```diff
@@ -0,0 +1,19 @@
+class StringCalculatorTest : StringSpec({
+    "문자열을 입력 받아 사칙 연산을 수행한다." {
+        val stringCalculator = StringCalculator(ExpressionFactory(OneSpaceSeparationStrategy()))
+        val parameters = listOf(
+            StringCalculatorFixture("2 + 3 * 4 / 2", 10),
+            StringCalculatorFixture("2 * 8 / 3 - 5", 0),
+            StringCalculatorFixture("3 + 4 - 10 * 8", -24),
+        )
+
+        parameters.forEach {
+            stringCalculator.calculate(it.expression) shouldBe it.expected
+        }
```
테스트하기 쉽도록 CalculatorFixtures 파일에 wrapping 클래스를 만들어 주셨네요 👍 
테스트 코드도 결국 코드이기 때문에, 테스트케이스가 추가될 떄마다 매번 data class가 생겨나게 되고, 결국 코드 관리가 번거로워질 것 같아요.
그리고 StringCalculatorFixture 객체를 생성하는 코드를 보았을 때, 이 것이 무엇을 의미하는지 잘 모르겠네요. 😭 

테스트코드가 로직을 검증하는 역할도 하지만, 그 과정도 중요합니다.
Wrapping 클래스를 사용하지 않고 테스트 코드를 작성해보면 어떨까요? (이하 관련 내용 동일)

```suggestion
        val expressionGroup = listOf(
            "2 + 3 * 4 / 2",
            "2 * 8 / 3 - 5",
            "3 + 4 - 10 * 8"
        )
        val expectedGroup = listOf(10, 0, -24)

        expressionGroup.forEachIndexed { index, expression ->
            stringCalculator.calculate(expression) shouldBe expectedGroup[index]
        }
```
혹시나 해서 예시를 적어두었습니다 😅

leeyohan93 replied:
각각 나누어서 변수에 할당하는 것도 좋은 방식이네요! 👍 
다만 생각이 드는 것이 Junit으로 구현하게 된다면 @CsvSource로 간단하게 구현가능할 것 같다는 생각도 드는데..
이건 제가 아직 Kotest를 제대로 활용하고 있지 않은 문제인 것 같네요 😂

BeokBeok reviewed (approved):
피드백 반영하시느라 고생하셨습니다. 👍 
추가로 고민해볼만한 의견들을 코맨트로 작성하였으니, 충분히 고민해보시고 도전해보세요. 💪 
다음 미션을 진행함과 동시에 작성된 코맨트들도 반영해주세요. 🙏

BeokBeok commented:
@leeyohan93 
추가로, 현재 ktlint가 통과되지 않고 있어요.
```
kotlin-racingcar/src/main/kotlin/calculator/ExpressionFactory.kt:46:21 Missing spacing before "{" (curly-spacing)
```
이 부분도 다음 미션에서 체크 부탁드려요 🙏
